test/recipes/82-test_ocsp_cert_chain.t

```diff
@@ -0,0 +1,147 @@
+    if alldisabled(available_protocols("tls"));
+plan skip_all => "$test_name is not available Windows or VMS"
+    if $^O =~ /^(VMS|MSWin32)$/;
+
+plan tests => 3;
+
+my $shlib_wrap   = bldtop_file("util", "shlib_wrap.sh");
+my $apps_openssl = bldtop_file("apps", "openssl");
+
+my $index_txt             = srctop_file("test", "ocsp-tests", "index.txt");
+my $ocsp_pem              = srctop_file("test", "ocsp-tests", "ocsp.pem");
+my $intermediate_cert_pem = srctop_file("test", "ocsp-tests", "intermediate-cert.pem");
+
+my $server_pem            = srctop_file("test", "ocsp-tests", "server.pem");
+
+# return random integer r such that A <= r < B
+sub randrange {
+    my ($A, $B) = @_;
+    die "invalid range" unless $A < $B;
+    my $r = $A + int(rand($B - $A));
+    return $r;
+}
+
+sub run_test {
+
+    # this test starts two servers that listen on respective ports.
+    # that can be problematic since the ports may not be available
+    # (e.g. when multiple instances of the test are run on the same
+    # machine).
+
+    # we use random ports to help avoid this.
+
+    # using a random port with s_server is straightforward.  doing so
+    # with the ocsp responder required some investigation because the
+    # url for the ocsp responder is usually included in the server's
+    # cert (normally, in the authority information access extension,
+    # and it would be complicated to change that when the test
+    # executes).  however, s_server has an option "-status_url" that
+    # can be used to specify a fallback url when no url is specified
+    # in the cert.  that is what we do here.
+
+    my $ocsp_port = randrange(19200, 19300);
+    my $server_port = randrange(4400, 4500);
+
+    # openssl ocsp -port 19254 -index index.txt -rsigner ocsp.pem -CA intermediate-cert.pem
+    my @ocsp_cmd = ("ocsp", "-port", "$ocsp_port", "-index", $index_txt, "-rsigner", $ocsp_pem, "-CA", $intermediate_cert_pem);
+    my $ocsp_pid = open3(my $ocsp_i, my $ocsp_o, my $ocsp_e = gensym, $shlib_wrap, $apps_openssl, @ocsp_cmd);
+
+    ## ipv4
+    # ACCEPT 0.0.0.0:19254 PID=620007
+    ## ipv6
+    # ACCEPT [::]:19254 PID=620007
+    my $port = "0";
+    while (<$ocsp_o>) {
+        print($_);
+        chomp;
+        if (/^ACCEPT 0.0.0.0:(\d+)/) {
+            $port = $1;
+            last;
+        } elsif (/^ACCEPT \[::\]:(\d+)/) {
+            $port = $1;
+            last;
+        } else {
+            last;
+        }
+    }
+    ok($port eq $ocsp_port, "ocsp server port check");
+
+    print("ocsp server ready, listening on port $port\n");
+
+    # openssl s_server -accept 4433 -cert server.pem -cert_chain intermediate-cert.pem \
+    #                  -status_verbose -status_url http://127.0.0.1:19254/ocsp
+    my @s_server_cmd = ("s_server", "-accept", "$server_port", "-cert", $server_pem, "-cert_chain", $intermediate_cert_pem,
+                        "-status_verbose", "-status_url", "http://127.0.0.1:${ocsp_port}/ocsp");
```

Comment:
This is going to cause problems. Some machines that we build on do not have IPv4 available!!

Reply:
So on those machines, the loopback interface does not recognize 127.0.0.1?  Is that correct?

Reply:
Right